src/app/utils/socket-utils.ts

```diff
@@ -254,6 +245,172 @@ export function isModCostVisible(plug: PluggableInventoryItemDefinition): boolea
   );
 }
 
+const ARMOR_STAT_CATEGORYSTYLE = 2251952357;
+
+function filterSocketCategories(
+  categories: DimSocketCategory[],
+  sockets: DimSockets,
+  allowCategory: (cat: DimSocketCategory) => boolean,
+  allowSocket: (socket: DimSocket) => boolean
+): { categories: DimSocketCategory[]; socketsByCategory: Map<DimSocketCategory, DimSocket[]> } {
+  // Pre-calculate the list of sockets we'll display for each category
+  const socketsByCategory = new Map<DimSocketCategory, DimSocket[]>();
+  for (const category of categories) {
+    if (!allowCategory(category)) {
+      continue;
+    }
+    const categorySockets = getSocketsByIndexes(sockets, category.socketIndexes).filter(
+      (socketInfo) => socketInfo.plugged?.plugDef.displayProperties.name && allowSocket(socketInfo)
+    );
+    if (categorySockets.length) {
+      socketsByCategory.set(category, categorySockets);
```

Comment:
Since maps remember the order that keys were inserted, do you need the category list alongside it at all?